### PR TITLE
Proposal: Add build support on a random node

### LINKED ISSUE #455: Add docker build
I'm working on service that allows users to create developer environments. I'm planning to use swarm instead of (or with) self written tool for load balancing.
User should be able to create dev environment by building container from Dockerfile.

### PR DESCRIPTION
**Pull request by vieux**
Fix #455 
Related to #602

```
docker build -t myimage 
docker run -e affinity:image==myimage myimage
```

### CODE REVIEW

**Comment by garagatyi** [comment]
I think that build mostly will be used to create containers, not pushing it somewhere. Thats why it doesn't make sense to create it on overloaded nodes, where containers can't be started.

**Comment by vieux** [comment]
@garagatyi yes, you can still do 
```
docker build -t user/myimage 
docker push user/myimage
docker run user/myimage ...
```

**Comment by vieux** [comment]
following @aluzzardi idea, "random" nos uses the filters and strategy selected. build will likely select the best node that match your needs.

**Comment by aluzzardi** [comment]
LGTM

**Comment by aluzzardi** [comment]
@garagatyi: How about now?

File: cluster/swarm/cluster.go [hunk 1]
```diff
@@ -295,3 +295,15 @@ func (c *Cluster) Info() [][2]string {
 
 	return info
 }
+
+// RANDOMENGINE returns a random engine.
+func (c *Cluster) RANDOMENGINE() (*cluster.Engine, error) {
```
Comment (abronan) [inline cluster/swarm/cluster.go]: Nit-picking on the name but why keeping **random*** if it defaults to the strategy used? (can be something like *PickEngine* or else?)

**Comment by abronan** [comment]
@aluzzardi way better than random :D

**Comment by aluzzardi** [comment]
@vieux: If possible, would it be too much of a hassle to add an integration test?

Eventually, before 0.3, it would be awesome to have an integration test for every single API endpoint so we can confidently say "we are 100% compatible with the Docker API".

Creating them as we add API endpoints will be less of a pain later :)

**Comment by garagatyi** [comment]
@vieux thanks for the improvement
@aluzzardi LGTM now

**Comment by vieux** [comment]
ping @aluzzardi

**Comment by aluzzardi** [comment]
@vieux I thought you were adding integration tests for this one before the review?

**Comment by aluzzardi** [comment]
Also, there's one commit that doesn't seem to pass tests (6961b54) - could you squash it with something else?

Ideally every single commit should build / pass tests. It will make our lives much easier should we need to git bisect someday.

**Comment by vieux** [comment]
@aluzzardi PTAL

File: cluster/engine.go [hunk 2]
```diff
@@ -156,7 +156,7 @@ func (e *Engine) RemoveImage(image *Image) ([]*dockerclient.ImageDelete, error)
 }
 
 // Refresh the list of images on the engine.
```
Comment (aluzzardi) [inline cluster/engine.go]: golint is complaining about this

**Comment by aluzzardi** [comment]
Any way to know whether the image was built "manually"?

I'm wondering if `run` could work without the need of affinity - otherwise, we won't be "compatible" with the docker API since `docker build && docker run` will fail (CI system for example).

**Comment by vieux** [comment]
rebased

**Comment by vieux** [comment]
@aluzzardi as we discussed it should be fixed in docker itself, I'll open another issue.

**Comment by aluzzardi** [comment]
Yep, looks good!

Last item: Could you please create an issue to refactor `build`?

Ideally we would add a `cluster.Build` function (along with `Engine.Build` and support in `dockerclient`), in the same way it was done for `load`: https://github.com/docker/swarm/pull/609

**Pull request merged**